src/topp/MascotAdapterOnline.cpp

```diff
@@ -289,124 +297,187 @@ class TOPPMascotAdapterOnline :
+          }
+          catch (Exception::ElementNotFound&)
+          {
+            OPENMS_LOG_ERROR << "Error: Failed to look up spectrum native ID for peptide identification with retention time '" + String(pep.getRT()) + "'." << endl;
+          }
+        }
+
+        if (internal_decoys)
+        {
+          vector<PeptideIdentification> decoy_pep_ids;
+          ProteinIdentification decoy_prot_id;
+          parseMascotResponse_(current_batch, true, mascot_query, decoy_prot_id, decoy_pep_ids);  // decoys
+
+          // reannotate proper spectrum native id if missing
+          for (auto& pep : decoy_pep_ids)
+          {
+            // no need to reannotate
+            if (pep.metaValueExists("spectrum_reference") 
+              && !(static_cast<String>(pep.getMetaValue("spectrum_reference")).empty()))
+            {
+              continue;
+            } 
+
+            try
+            { 
+              Size index = lookup.findByRT(pep.getRT());
+              pep.setMetaValue("spectrum_reference", exp[index].getNativeID());
+            }
+            catch (Exception::ElementNotFound&)
+            {
+              OPENMS_LOG_ERROR << "Error: Failed to look up spectrum native ID for peptide identification with retention time '" + String(pep.getRT()) + "'." << endl;
+            }
+          }
+          mergeIDs_(prot_id, decoy_prot_id, pep_ids, decoy_pep_ids);
         }
-        // remove proteins
-        prot_id.getHits().clear();
       }
-    }
 
-    String search_number = mascot_query->getSearchIdentifier();
-    if (search_number.empty())
-    {
-      writeLog_("Error: Failed to extract the Mascot search identifier (search number).");
-      if (mascot_query_param.exists("skip_export") &&
-          mascot_query_param.getValue("skip_export").toBool())
+      String search_number = mascot_query->getSearchIdentifier();
+      if (search_number.empty())
       {
-        return PARSE_ERROR;
+        writeLog_("Error: Failed to extract the Mascot search identifier (search number).");
+        if (mascot_query_param.exists("skip_export") &&
+            mascot_query_param.getValue("skip_export").toBool())
+        {
+          return PARSE_ERROR;
+        }
+      }
+      else 
+      {
+        prot_id.setMetaValue("SearchNumber", search_number);
       }
-    }
-    else prot_id.setMetaValue("SearchNumber", search_number);
 
-    // clean up
-    delete mascot_query;
+      // clean up
+      delete mascot_query;
+      
+      current_batch.clear(true); // clear meta data
+
+      mergeIDs_(all_prot_id, prot_id, all_pep_ids, pep_ids);
+    }
 
     //-------------------------------------------------------------
     // writing output
     //-------------------------------------------------------------
+    all_prot_id.setPrimaryMSRunPath({ in }, exp);
+
+    // remove proteins as protein links seem are broken and reindexing is needed
+    all_prot_id.getHits().clear(); 
+    all_prot_ids.push_back(all_prot_id);
```

Comment:
this part is a bit confusing to me, but I guess it was a work around

Reply:
maybe renaming the variables `all_prot_id` and `all_prot_ids` to distinguish them more clearly could help?